includes/panopto_derivatives.inc

```diff
@@ -5,37 +5,356 @@
     }
   }
 }
 
+/*
+ * Requests token and returns token
+ * @param String $oauthendpoint 
+ */
+function islandora_video_streaming_panopto_get_token($oauthendpoint) {
+  $client_id = variable_get('islandora_video_streaming_panopto_client_id');
+  $secret_client_id = variable_get('islandora_video_streaming_panopto_secret_client_id');
+  $username = variable_get('islandora_video_streaming_panopto_username');
+  $password = variable_get('islandora_video_streaming_panopto_password');
+  
+  $response = null;
+  $token = null;
+  
+  if (isset($client_id) && isset($secret_client_id)) {
+    $authorization = base64_encode("$client_id:$secret_client_id");
+    $headers = array("Authorization: Basic {$authorization}", "Content-Type: application/x-www-form-urlencoded");
+    $content = 'grant_type=password&username='.urlencode(strtolower($username)).'&password='.urlencode($password).'&scope=api';
+    $response = islandora_video_streaming_panopto_request_curl($oauthendpoint, $headers, $content, 'POST', array(200 => 'access_token'));
+  }
+  if ($response) {
+    $token = $response['access_token'];
+  }
+  return $token;
+}
+
+/*
+ * Retrieves session needed for upload
+ * @param $token
+ * @param $uploadendpoint
+ */
+function islandora_video_streaming_panopto_get_session($token, $uploadendpoint) {
+  $session = null;
+  $headers = array("Authorization: Bearer {$token}", "Content-Type: application/json");
+  $folder_id = variable_get('islandora_video_streaming_panopto_folder_id');
+  $content = json_encode(array(
+    'FolderId' => $folder_id,
+  ));
+  $response = islandora_video_streaming_panopto_request_curl($uploadendpoint, $headers, $content, 'POST', array(201 => 'ID'));
+  if ($response) {
+    $session = $response;
+  }
+  return $session;
+}
+
+/*
+ * @param $endpoint
+ * @param $headers
+ * @param $content
+ * @param $type
+ * @param $expect
+ */
+function islandora_video_streaming_panopto_request_curl($endpoint, $headers, $content = '', $type = 'GET', $expect = array()) {
+  $cURL = curl_init();
+  curl_setopt($cURL, CURLOPT_URL, $endpoint);
+  curl_setopt($cURL, CURLOPT_RETURNTRANSFER, true);
+  curl_setopt($cURL, CURLOPT_HEADER, 1);
+  curl_setopt($cURL, CURLINFO_HEADER_OUT, true);
+  switch (strtoupper($type)) {
+    case 'POST':
+    curl_setopt($cURL, CURLOPT_POST, true);
+    break;
+    case 'PUT':
+    curl_setopt($cURL, CURLOPT_CUSTOMREQUEST, 'PUT');
+    break;
+  }
+  if ($content) {
+    curl_setopt($cURL, CURLOPT_POSTFIELDS, $content);
+  }
+  curl_setopt($cURL, CURLOPT_HTTPHEADER, $headers);
+  $response = curl_exec($cURL);
+  $header_size = curl_getinfo($cURL, CURLINFO_HEADER_SIZE);
+  $response_code = curl_getinfo($cURL, CURLINFO_RESPONSE_CODE);
+  $headerSent = curl_getinfo($cURL, CURLINFO_HEADER_OUT );
+  curl_close($cURL);
+  if ($response === false) {
+    $messages = array(
+      'message' => t("Unable to receive response from curl_exec() for object @pid"),
+      'message_sub' => array('@pid' => $object->id),
```

Comment:
You don't have `$object` in context here.

Reply:
Fixed.